# Data management using Pandas

**Data management** is a crucial component to statistical analysis and data science work. 

This notebook will show you how to import, view, undertand, and manage your data using the [Pandas](http://pandas.pydata.org) data processing library, i.e., the notebook will demonstrates how to read a dataset into Python, and obtain a basic understanding of its content.

Note that **Python** by itself is a general-purpose programming language and does not provide high-level data processing capabilities.  The **Pandas** library was developed to meet this need. **Pandas** is the most popular Python library for data manipulation, and we will use it extensively in this course. **Pandas** provides high-performance, easy-to-use data structures and data analysis tools.

The main data structure that **Pandas** works with is called a **Data Frame**. This is a two-dimensional table of data in which the rows typically
represent cases and the columns represent variables (e.g. data used in this tutorial).  Pandas also has a one-dimensional data structure called a **Series** that we will encounter when accesing a single column of a Data Frame.

Pandas has a variety of functions named `read_xxx` for reading data in different formats.  Right now we will focus on reading `csv` files, which stands for comma-separated values. However the other file formats include `excel`, `json`, and `sql`. 

There are many other options to `read_csv` that are very useful.  For example, you would use the option `sep='\t'` instead of the default `sep=','` if the fields of your data file are delimited by tabs instead of commas.  See [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) for the full documentation for `read_csv`.


## Acknowledgments

- The dataset used in this tutorial is from https://www.coursera.org/ from the course "Understanding and Visualizing Data with Python" by University of Michigan


# Importing libraries


In [1]:
# Import the packages that we will be using
import pandas as pd

# Importing data

In [2]:
# Define the col names for the iris dataset
colnames = ["Sepal_Length", "Sepal_Width","Petal_Length","Petal_Width", "Flower_Name"]

# Dataset url
# from HHDD
url = "datasets/iris/iris.csv"

dataframe = pd.read_csv(url, header = None, names = colnames )

If we want to print the information about th output object type we would simply type the following: type(df)

In [3]:
type(dataframe)

pandas.core.frame.DataFrame

# Exploring the content of the data set

Use the `shape` method to determine the numbers of rows and columns in a data frame. This can be used to confirm that we have actually obtained the data the we are expecting.

Based on what we see below, the data set being read here has $N_r$ rows, corresponding to $N_r$ observations, and $N_c$ columns, corresponding to $N_c$ variables in this particular data file.

In [4]:
Nr, Nc= dataframe.shape
print("Number of rows:", Nr)
print("Number of columns:", Nc)

Number of rows: 150
Number of columns: 5


If we want to show the entire data frame we would simply write the following:

In [5]:
dataframe

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width     Flower_Name
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

[150 rows x 5 columns]

As you can see, we have a 2-Dimensional object where each row is an independent observation and each coloum is a variable.

Now, use the the `head()` function to show the first 5 rows of our data frame

In [6]:
dataframe.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

Also, you can use the the `tail()` function to show the last 5 rows of our data frame

In [7]:
dataframe.tail()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width     Flower_Name
145           6.7          3.0           5.2          2.3  Iris-virginica
146           6.3          2.5           5.0          1.9  Iris-virginica
147           6.5          3.0           5.2          2.0  Iris-virginica
148           6.2          3.4           5.4          2.3  Iris-virginica
149           5.9          3.0           5.1          1.8  Iris-virginica

The columns in a Pandas data frame have names, to see the names, use the `columns` method:

To gather more information regarding the data, we can view the column names with the following function:

In [8]:
dataframe.columns

Index(['Sepal_Length', 'Sepal_Width', 'Petal_Length', 'Petal_Width',
       'Flower_Name'],
      dtype='object')

Be aware that every variable in a Pandas data frame has a data type.  There are many different data types, but most commonly you will encounter floating point values (real numbers), integers, strings (text), and date/time values.  When Pandas reads a text/csv file, it guesses the data types based on what it sees in the first few rows of the data file.  Usually it selects an appropriate type, but occasionally it does not.  To confirm that the data types are consistent with what the variables represent, inspect the `dtypes` attribute of the data frame.

In [9]:
dataframe.dtypes

Sepal_Length    float64
Sepal_Width     float64
Petal_Length    float64
Petal_Width     float64
Flower_Name      object
dtype: object

Summary statistics, which include things like the mean, min, and max of the data, can be useful to get a feel for how large some of the variables are and what variables may be the most important. 

In [10]:
# Summary statistics for the quantitative variables
dataframe.describe()


Sepal_Length  Sepal_Width  Petal_Length  Petal_Width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [11]:
# Drop observations with NaN values
#df.Age.dropna().describe()
#df.Wingspan.dropna().describe()
dataframe.Sepal_Length.dropna().describe()

count    150.000000
mean       5.843333
std        0.828066
min        4.300000
25%        5.100000
50%        5.800000
75%        6.400000
max        7.900000
Name: Sepal_Length, dtype: float64

It is also possible to get statistics on the entire data frame or a column as follows

- `df.mean()` Returns the mean of all columns
- `df.corr()` Returns the correlation between columns in a data frame
- `df.count()` Returns the number of non-null values in each data frame column
- `df.max()` Returns the highest value in each column
- `df.min()` Returns the lowest value in each column
- `df.median()` Returns the median of each column
- `df.std()` Returns the standard deviation of each column

In [1]:
dataframe.Sepal_Length.mean()

NameError: name 'dataframe' is not defined

# How to write a data frame to a File

To save a file with your data simply use the `to_csv` attribute

Examples:
- df.to_csv('myDataFrame.csv')
- df.to_csv('myDataFrame.csv', sep='\t')

In [13]:
dataframe.to_csv('myIrisDataFrame.csv')

# Rename columns

To change the name of a colum use the `rename` attribute

Example:

df = df.rename(columns={"Age": "Edad"})

df.head()

In [14]:
dataframe = dataframe.rename(columns={"Sepal_Length": "Longitud del Sépalo"})
dataframe.head()

Longitud del Sépalo  Sepal_Width  Petal_Length  Petal_Width  Flower_Name
0                  5.1          3.5           1.4          0.2  Iris-setosa
1                  4.9          3.0           1.4          0.2  Iris-setosa
2                  4.7          3.2           1.3          0.2  Iris-setosa
3                  4.6          3.1           1.5          0.2  Iris-setosa
4                  5.0          3.6           1.4          0.2  Iris-setosa

In [15]:
# Back to the original name
dataframe = dataframe.rename(columns={"Longitud del Sépalo": "Sepal_Length"})
dataframe.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

# Selection of colums

As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent observations or samples and the columns represent variables.  One common manipulation of a data frame is to extract the data for one case or for one variable.  There are several ways to do this, as shown below.

To extract all the values for one column (variable), use one of the following alternatives.

In [16]:
#a = df.Age
#b = df["Age"]
#c = df.loc[:, "Age"]
#d = df.iloc[:, 1]

a = dataframe.Sepal_Length
b = dataframe["Sepal_Length"]
c = dataframe.loc[:, "Sepal_Length"]
d = dataframe.iloc[:, 0]

#print(d)
print(a)
print(b)
print(c)
print(d)

#df[["Gender", "GenderGroup"]]
dataframe[["Flower_Name", "Sepal_Length", "Sepal_Width"]]

0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: Sepal_Length, Length: 150, dtype: float64
0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: Sepal_Length, Length: 150, dtype: float64
0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: Sepal_Length, Length: 150, dtype: float64
0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: Sepal_Length, Length: 150, dtype: float64


Flower_Name  Sepal_Length  Sepal_Width
0       Iris-setosa           5.1          3.5
1       Iris-setosa           4.9          3.0
2       Iris-setosa           4.7          3.2
3       Iris-setosa           4.6          3.1
4       Iris-setosa           5.0          3.6
..              ...           ...          ...
145  Iris-virginica           6.7          3.0
146  Iris-virginica           6.3          2.5
147  Iris-virginica           6.5          3.0
148  Iris-virginica           6.2          3.4
149  Iris-virginica           5.9          3.0

[150 rows x 3 columns]

# Slicing a data set

As discussed above, a Pandas data frame is a rectangular data table, in which the rows represent cases and the columns represent variables.  One common manipulation of a data frame is to extract the data for one observation or for one variable.  There are several ways to do this, as shown below.

Lets say we would like to splice our data frame and select only specific portions of our data.  There are three different ways of doing so.

1. .loc()
2. .iloc()
3. .ix()

We will cover the .loc() and .iloc() splicing functions.


The attibute **.loc()** uses labels/column names, in specific, it takes two single/list/range operator separated by ',', the first one indicates the rows and the second one indicates columns.

In [17]:
# Return all observations of CWDistance
#df.loc[:,"CWDistance"]
dataframe.loc[:,"Flower_Name"]

# Return a subset of observations of CWDistance
#df.loc[:9, "CWDistance"]
dataframe.loc[:9, "Flower_Name"]

# Select all rows for multiple columns, ["Gender", "GenderGroup"]
#df.loc[:,["Gender", "GenderGroup"]]
dataframe.loc[:,["Sepal_Length", "Flower_Name"]]

# Select multiple columns, ["Gender", "GenderGroup"]me 
#keep = ['Gender', 'GenderGroup']
#df_gender = df[keep]
keep = ['Sepal_Length', 'Flower_Name']
df_gender = dataframe[keep]

# Select few rows for multiple columns, ["CWDistance", "Height", "Wingspan"]
#df.loc[4:9, ["CWDistance", "Height", "Wingspan"]]
dataframe.loc[4:9, ["Flower_Name", "Sepal_Length", "Sepal_Width"]]

# Select range of rows for all columns
#df.loc[10:15,:]
dataframe.loc[10:15,:]


Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name
10           5.4          3.7           1.5          0.2  Iris-setosa
11           4.8          3.4           1.6          0.2  Iris-setosa
12           4.8          3.0           1.4          0.1  Iris-setosa
13           4.3          3.0           1.1          0.1  Iris-setosa
14           5.8          4.0           1.2          0.2  Iris-setosa
15           5.7          4.4           1.5          0.4  Iris-setosa

The attribute **iloc()** is an integer based slicing.

In [18]:
# .
#df.iloc[:, :4]
dataframe.iloc[:, :4]

# .
#df.iloc[:4, :]
dataframe.iloc[:4, :]

# .
#df.iloc[:, 3:7]
dataframe.iloc[:, 3:7]

# .
#df.iloc[4:8, 2:4]
dataframe.iloc[4:8, 2:4]

# This is incorrect:
#df.iloc[1:5, ["Flower_Name", "Sepal_Length", "Sepal_Width"]]

Petal_Length  Petal_Width
4           1.4          0.2
5           1.7          0.4
6           1.4          0.3
7           1.5          0.2

# Get unique existing values

List unique values in the one of the columns

df.Gender.unique()


In [19]:
# List unique values in the df['Flower_Name'] column
dataframe.Flower_Name.unique()

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

In [20]:
# Lets explore df["Sepal_Length] as well
dataframe.Sepal_Length.unique()

array([5.1, 4.9, 4.7, 4.6, 5. , 5.4, 4.4, 4.8, 4.3, 5.8, 5.7, 5.2, 5.5,
       4.5, 5.3, 7. , 6.4, 6.9, 6.5, 6.3, 6.6, 5.9, 6. , 6.1, 5.6, 6.7,
       6.2, 6.8, 7.1, 7.6, 7.3, 7.2, 7.7, 7.4, 7.9])

# Filter, Sort and Groupby



With **Filter** you can use different conditions to filter columns. For example, df[df[year] > 1984] would give you only the column year is greater than 1984. You can use & (and) or | (or) to add different conditions to your filtering. This is also called boolean filtering.

df[df["Height"] >= 70]

With **Sort** is possible to sort values in a certain column in an ascending order using `df.sort_values("ColumnName")` or in descending order using `df.sort_values(ColumnName, ascending=False)`. 

Furthermore, it’s possible to sort values by Column1Name in ascending order then Column2Name in descending order by using `df.sort_values([Column1Name,Column2Name],ascending=[True,False])`


df.sort_values("Height") 
#df.sort_values("Height",ascending=False)

The attribute **Groupby** involves splitting the data into groups based on some criteria, applying a function to each group independently and combining the results into a data structure. df.groupby(col) returns a groupby object for values from one column while df.groupby([col1,col2]) returns a groupby object for values from multiple columns.

df.groupby(['Gender'])

Size of each group

df.groupby(['Gender']).size()

df.groupby(['Gender','GenderGroup']).size()

This output indicates that we have two types of combinations. 

* Case 1: Gender = F & Gender Group = 1 
* Case 2: Gender = M & GenderGroup = 2.  

This validates our initial assumption that these two fields essentially portray the same information.

# Data Cleaning: handle with missing data

Before getting started to work with your data, it's a good practice to observe it thoroughly to identify missing values and handle them accordingly.

When reading a dataset using Pandas, there is a set of values including 'NA', 'NULL', and 'NaN' that are taken by default to represent a missing value.  The full list of default missing value codes is in the '`read_csv`' documentation [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html).  This document also explains how to change the way that '`read_csv`' decides whether a variable's value is missing.

Pandas has functions called `isnull` and `notnull` that can be used to identify where the missing and non-missing values are located in a data frame.  

Below we use these functions to count the number of missing and non-missing values in each variable of the datasetr.

In [21]:
dataframe.isnull()


Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name
0           False        False         False        False        False
1           False        False         False        False        False
2           False        False         False        False        False
3           False        False         False        False        False
4           False        False         False        False        False
..            ...          ...           ...          ...          ...
145         False        False         False        False        False
146         False        False         False        False        False
147         False        False         False        False        False
148         False        False         False        False        False
149         False        False         False        False        False

[150 rows x 5 columns]

Unfortunately, our output indicates that some of our columns contain missing values so we are no able to continue on doing analysis with those colums

In [22]:
dataframe.notnull()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name
0            True         True          True         True         True
1            True         True          True         True         True
2            True         True          True         True         True
3            True         True          True         True         True
4            True         True          True         True         True
..            ...          ...           ...          ...          ...
145          True         True          True         True         True
146          True         True          True         True         True
147          True         True          True         True         True
148          True         True          True         True         True
149          True         True          True         True         True

[150 rows x 5 columns]

In [23]:
#df.isnull().sum()
#df.notnull().sum()
dataframe.isnull().sum()

Sepal_Length    0
Sepal_Width     0
Petal_Length    0
Petal_Width     0
Flower_Name     0
dtype: int64

In [24]:
dataframe.notnull().sum()

Sepal_Length    150
Sepal_Width     150
Petal_Length    150
Petal_Width     150
Flower_Name     150
dtype: int64

Now we use these functions to count the number of missing and non-missing values in a single variable in the dataset

print( df.Height.notnull().sum() )

print( pd.isnull(df.Height).sum() )

In [25]:
print( dataframe.Flower_Name.notnull().sum() )

print( pd.isnull(dataframe.Sepal_Width).sum() )

150
0


In [26]:
# Extract all non-missing values of one of the columns into a new variable
x = dataframe.Flower_Name.dropna().describe()
x.describe()

count                   4
unique                  4
top       Iris-versicolor
freq                    1
Name: Flower_Name, dtype: object

# Add and eliminate columns

In some cases it is useful to create or eiminate new columns 

In [27]:
#df.head()
dataframe.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [28]:
# Add a new column with new data

# Create a column data
#NewColumnData = df.Age/df.Age
NewColumnData = dataframe.Sepal_Length*2.54/100

# Insert that column in the data frame
#df.insert(12, "ColumnInserted", NewColumnData, True)
dataframe.insert(5, "Longitud de Sépalo en metros", NewColumnData, True)

#df.head()
dataframe.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Longitud de Sépalo en metros  
0                       0.12954  
1                       0.12446  
2                       0.11938  
3                       0.11684  
4                       0.12700

In [29]:
# # Eliminate inserted column
# df.drop("ColumnInserted", axis=1, inplace = True)
dataframe.drop("Longitud de Sépalo en metros", axis=1, inplace = True)

# #df.drop(columns=['ColumnInserted'], inplace = True)
# # Remove three columns as index base
# #df.drop(df.columns[[12]], axis = 1, inplace = True)
#
# df.head()
dataframe.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [30]:
# # Add new column derived from existing columns
#
# # The new column is a function of another column
# df["AgeInMonths"] = df["Age"] * 12
dataframe["Ancho de Sépalo en metros"] = dataframe["Sepal_Width"]*2.54/100
dataframe.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   Ancho de Sépalo en metros  
0                    0.08890  
1                    0.07620  
2                    0.08128  
3                    0.07874  
4                    0.09144

In [31]:
# # Eliminate inserted column
# df.drop("AgeInMonths", axis=1, inplace = True)
dataframe.drop("Ancho de Sépalo en metros", axis=1, inplace = True)
#
dataframe.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [32]:
# Add a new column with text labels reflecting the code's meaning

# df["GenderGroupNew"] = df.GenderGroup.replace({1: "Female", 2: "Male"})
dataframe["FlowerNameGroup"] = dataframe.Flower_Name.replace({"Iris-setosa": 1, "Iris-versicolor": 2, "Iris-virginica": 3})
# Show the first 5 rows of the created data frame
dataframe.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name  \
0           5.1          3.5           1.4          0.2  Iris-setosa   
1           4.9          3.0           1.4          0.2  Iris-setosa   
2           4.7          3.2           1.3          0.2  Iris-setosa   
3           4.6          3.1           1.5          0.2  Iris-setosa   
4           5.0          3.6           1.4          0.2  Iris-setosa   

   FlowerNameGroup  
0                1  
1                1  
2                1  
3                1  
4                1

In [33]:
dataframe.tail()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width     Flower_Name  \
145           6.7          3.0           5.2          2.3  Iris-virginica   
146           6.3          2.5           5.0          1.9  Iris-virginica   
147           6.5          3.0           5.2          2.0  Iris-virginica   
148           6.2          3.4           5.4          2.3  Iris-virginica   
149           5.9          3.0           5.1          1.8  Iris-virginica   

     FlowerNameGroup  
145                3  
146                3  
147                3  
148                3  
149                3

In [34]:
## Eliminate inserted column
# df.drop("GenderGroupNew", axis=1, inplace = True)
dataframe.drop("FlowerNameGroup", axis=1, inplace = True)
##df.drop(['GenderGroupNew'],vaxis='columns',vinplace=True)
dataframe.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [35]:
## Add a new column with strata based on these cut points
#
## Create a column data
#NewColumnData = df.Age/df.Age
NewColumnData = dataframe.Sepal_Length/dataframe.Sepal_Length
#
## Insert that column in the data frame
#df.insert(1, "ColumnStrata", NewColumnData, True)
dataframe.insert(1, "ColumnStrata", NewColumnData, True)

#df["ColumnStrata"] = pd.cut(df.Height, [60., 63., 66., 69., 72., 75., 78.])
dataframe["ColumnStrata"] = pd.cut(dataframe.Sepal_Length, [1., 2., 3., 4., 5., 6., 7.])

## Show the first 5 rows of the created data frame
dataframe.head()




Sepal_Length ColumnStrata  Sepal_Width  Petal_Length  Petal_Width  \
0           5.1   (5.0, 6.0]          3.5           1.4          0.2   
1           4.9   (4.0, 5.0]          3.0           1.4          0.2   
2           4.7   (4.0, 5.0]          3.2           1.3          0.2   
3           4.6   (4.0, 5.0]          3.1           1.5          0.2   
4           5.0   (4.0, 5.0]          3.6           1.4          0.2   

   Flower_Name  
0  Iris-setosa  
1  Iris-setosa  
2  Iris-setosa  
3  Iris-setosa  
4  Iris-setosa

In [36]:
## Eliminate inserted column
#df.drop("ColumnStrata", axis=1, inplace = True)
dataframe.drop("ColumnStrata", axis=1, inplace = True)
#
#df.head()
dataframe.head()




Sepal_Length  Sepal_Width  Petal_Length  Petal_Width  Flower_Name
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [37]:
# Drop several "unused" columns
#vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
vars = ["Petal_Length", "Petal_Width"]
#df.drop(vars, axis=1, inplace = True)

dataframe.drop(vars, axis=1, inplace = True)
dataframe.head()

Sepal_Length  Sepal_Width  Flower_Name
0           5.1          3.5  Iris-setosa
1           4.9          3.0  Iris-setosa
2           4.7          3.2  Iris-setosa
3           4.6          3.1  Iris-setosa
4           5.0          3.6  Iris-setosa

In [38]:
#Rewrite the original dataframe
dataframe = pd.read_csv(url, header = None, names = colnames )

In [39]:
# Drop several "unused" columns
#vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
vars = ["Sepal_Length", "Sepal_Width"]
#df.drop(vars, axis=1, inplace = True)

dataframe.drop(vars, axis=1, inplace = True)
dataframe.head()

Petal_Length  Petal_Width  Flower_Name
0           1.4          0.2  Iris-setosa
1           1.4          0.2  Iris-setosa
2           1.3          0.2  Iris-setosa
3           1.5          0.2  Iris-setosa
4           1.4          0.2  Iris-setosa

In [40]:
#Rewrite the original dataframe
dataframe = pd.read_csv(url, header = None, names = colnames )

In [41]:
# Drop several "unused" columns
dataframe["FlowerNameGroup"] = dataframe.Flower_Name.replace({"Iris-setosa": 1, "Iris-versicolor": 2, "Iris-virginica": 3})
dataframe.drop("Flower_Name", axis=1, inplace = True)
vars = ["Petal_Length", "Petal_Width"]
#df.drop(vars, axis=1, inplace = True)
dataframe.drop(vars, axis=1, inplace = True)
dataframe.head()

Sepal_Length  Sepal_Width  FlowerNameGroup
0           5.1          3.5                1
1           4.9          3.0                1
2           4.7          3.2                1
3           4.6          3.1                1
4           5.0          3.6                1

# Add and eliminate rows

In some cases it is requiered to add new observations (rows) to the data set

In [42]:
# Print tail


In [43]:
#df.loc[len(df.index)] = [26, 24, 'F', 1, 'Y', 1, 66, 'NaN', 68, 'N', 0, 3] 
#
#df.tail()




In [44]:
## Eliminate inserted row
#df.drop([28], inplace = True )
#
#df.tail()




# Cleaning your data: drop out unused columns and/or drop out rows with any missing values

In [45]:
# Drop unused columns
#vars = ["ID", "GenderGroup", "GlassesGroup", "CompleteGroup"]
#df.drop(vars, axis=1, inplace = True)

#vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
#df = df[vars]

# Drop rows with any missing values
#df = df.dropna()

# Drop unused columns and drop rows with any missing values
#vars = ["Age", "Gender", "Glasses", "Height", "Wingspan", "CWDistance", "Complete", "Score"]
#df = df[vars].dropna()

#df


# Final remarks


- The understanding of your dataset is essential
    - Number of observations
    - Variables
    - Data types: numerical or categorial
    - What are my variables of interest

- There are several ways to do the same thing

- Cleaning your dataset (dropping out rows with any missing values) is a good practice

- The **Pandas** library provides fancy, high-performance, easy-to-use data structures and data analysis tools


# Activity: work with the iris dataset

Repeat this tutorial with the iris data set and respond to the following inquiries

1. Calculate the statistical summary for each quantitative variables. Explain the results
    - Identify the name of each column
    "Sepal_Length", "Sepal_Width","Petal_Length","Petal_Width", "Flower_Name"
    - Identify the type of each column
    float64,float64,float64,float64,object
    - Minimum, maximum, mean, average, median, standar deviation
    
    Sepal_Length	Sepal_Width	Petal_Length	Petal_Width
    count	150.000000	150.000000	150.000000	150.000000
    mean	5.843333	3.057333	3.758000	1.199333
    std	    0.828066	0.435866	1.765298	0.762238
    min	    4.300000	2.000000	1.000000	0.100000
    25%	    5.100000	2.800000	1.600000	0.300000
    50%	    5.800000	3.000000	4.350000	1.300000
    75%	    6.400000	3.300000	5.100000	1.800000
    max	    7.900000	4.400000	6.900000	2.500000
    
2. Are there missing data? If so, create a new dataset containing only the rows with the non-missing data
#No, there's no missing data

3. Create a new dataset containing only the petal width and length and the type of Flower
#Ok
4. Create a new dataset containing only the setal width and length and the type of Flower
#Ok
5. Create a new dataset containing the setal width and length and the type of Flower encoded as a categorical numerical column 
#Ok